### DISCUSSION TITLE
Simulation of Voltage-Driven Coils

### DISCUSSION THREAD

**Discussion by paul-hilscher** [mfem/mfem#3428] [Q&A]
Dear MFEM team, 

is there a way to model a voltage driven (electromagnetic) stranded coil in mfem ? 

The weak-forms in the coil in the domain $\Omega$ are given by:

1) The field problem in the a-formulation

```math
\left( \nu \, \mathrm{curl} \times \mathbf{a}, \mathrm{curl} \times \mathbf{a}' \right) _\Omega +
 I \left( \mathbf{\tau}, \mathbf{a}'  \right) _\Omega = 0  \quad,
```
where $\mathbf{a}$ is the magnetic vector potential, $I$ the (**unknown**) current flowing through each strand of the coil, and $\tau$ is the electric current density in the coil when a current of 1A is applied.

2) The electric circuit problem (which introduces a single additional dof)

```math
\int_\Omega \partial_t \mathbf{a} \cdot \mathbf{\tau} d\Omega + R I = V \quad,
```
where $R$ is the specified resistance and $V$ is the specified voltage.

Thanks a lot for any reply and many thanks for an amazing library.

**Comment by mlstowell** [comment]
Hi, @paul-hilscher,

This is not a class of problems that I am familiar with but I think MFEM could be used to solve them. A quick google search turns up an article which you may be familiar with: [Voltage-driven coils in finite-element formulations using a current vector and a magnetic scalar potential](https://www.researchgate.net/publication/224747203_Voltage-driven_coils_in_finite-element_formulations_using_a_current_vector_and_a_magnetic_scalar_potential). Their section 2.2 seems to describe the same type of problem that you outline above.

From looking at your description and the paper I think this could be implemented as a block system of size `(e+1)x(e+1)` where `e` is the number of edges in the mesh of your coil (I'm not sure how to handle multiple strands). One `e x e` sized block would be a curl-curl operator (as in MFEM's example 3) with the material coefficient $\nu$. The `1 x e` and `e x 1` blocks would need to be custom-built operators based on `LinearForm` objects (we've done this sort of thing before but we may not have any examples in the code yet). For a similar type of operator see issue #2818 and the prototype implementation in PR #3435.

If this sounds like a feasible approach I'd be happy to offer further advice if needed.

Best wishes,
Mark

**Reply by paul-hilscher** [reply]
Hi @mlstowell,

thanks a lot for your reply and pointers on how to move forward. I will give this a try and report back in a couple of days.
